Added swagger ui deployment.

immetoo opened the pull request.

johnament commented on swagger-ui/src/main/java/ws/ament/hammock/swagger/ui/SwaggerUIProvider.java:
```diff
@@ -0,0 +1,69 @@
+package ws.ament.hammock.swagger.ui;
```
Missing header.  Since you're writing most of this yourself, do you want to add your name to the NOTICES file?

immetoo replied:
Done

johnament commented on swagger-ui/pom.xml:
```diff
@@ -0,0 +1,90 @@
+            <artifactId>rest-assured</artifactId>
+        </dependency>
+    </dependencies>
+    <profiles>
```
Since this is only using servlets, don't need jersey profile.  Take a look at how defined at https://github.com/hammock-project/hammock/blob/master/rest-resteasy/pom.xml#L69-L106 - not sure if there's a cleaner way to make it inherit (maybe a new parent for all modules that are just servlet based?)

immetoo replied:
Changed to web based profiles, lets to cleaner in other issue.

johnament commented on swagger-ui/pom.xml:
```diff
@@ -0,0 +1,90 @@
+    </parent>
+    <artifactId>swagger-ui</artifactId>
+    <dependencies>
+        <dependency>
```
Are you expecting a user overrides this?  I see there's a config property for it, but a user also has to include the transitive dependency?

immetoo replied:
Had to have that property anyway for syncing the packaged version to runtime info.
This because all webjars have there version in the resource path, so multiple versions don't conflict.
Have now added scanning bean so artifact ui version can be updated without extra steps.

johnament commented on swagger-ui/src/main/java/ws/ament/hammock/swagger/ui/SwaggerUIConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+
+    public String getSwaggerUIVersion() {
+        if (swaggerUIVersion.isEmpty()) {
+            swaggerUIVersion = SwaggerUIPackagedVersion.find();
```
Interesting approach.  Why isn't this a CDI bean though?  Otherwise you can make it package visibility.

johnament commented:
With the updated profiles in place, you're getting build errors in tomcat.

```
[[1;31mERROR[m] [1;31mTests [0;1mrun: [0;1m2[m, [1;31mFailures: [0;1;31m1[m, Errors: 0, Skipped: 0, Time elapsed: 6.277 s[1;31m <<< FAILURE![m - in ws.ament.hammock.swagger.ui.[1mSwaggerUITest[m
[[1;31mERROR[m] testUILoaded(ws.ament.hammock.swagger.ui.SwaggerUITest)  Time elapsed: 0.646 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[Swagger UI]> but was:<[HTTP Status 501 – Not Implemented]>
	at ws.ament.hammock.swagger.ui.SwaggerUITest.testUILoaded(SwaggerUITest.java:65)

[[1;31mERROR[m] [1;31mFailures: [m
[[1;31mERROR[m] [1;31m  SwaggerUITest.testUILoaded:65 expected:<[Swagger UI]> but was:<[HTTP Status 501 – Not Implemented]>[m
[[1;31mERROR[m] [1;31mTests run: 2, Failures: 1, Errors: 0, Skipped: 0[m
[[1;31mERROR[m] Failed to execute goal [32morg.apache.maven.plugins:maven-surefire-plugin:2.20.1:test[m [1m(default-test)[m on project [36mswagger-ui[m: [1;31mThere are test failures.[m
[[1;31mERROR[m] [1;31m[m
[[1;31mERROR[m] [1;31mPlease refer to /home/travis/build/hammock-project/hammock/swagger-ui/target/surefire-reports for the individual test results.[m
[[1;31mERROR[m] [1;31mPlease refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.[m
[[1;31mERROR[m] -> [1m[Help 1][m
[[1;31mERROR[m] 
[[1;31mERROR[m] To see the full stack trace of the errors, re-run Maven with the [1m-e[m switch.
[[1;31mERROR[m] Re-run Maven using the [1m-X[m switch to enable full debug logging.
[[1;31mERROR[m] 
[[1;31mERROR[m] For more information about the errors and possible solutions, please read the following articles:
[[1;31mERROR[m] [1m[Help 1][m http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[[1;31mERROR[m] 
[[1;31mERROR[m] After correcting the problems, you can resume the build with the command
[[1;31mERROR[m]   [1mmvn <goals> -rf :swagger-ui[m

travis_time:end:10fecb70:start=1520659206922645582,finish=1520659409767178227,duration=202844532645
[0K
[31;1mThe command "mvn -q install -PWeld3 -Ptomcat-test" exited with 1.[0m
```

immetoo commented:
That was not easy, have done;
- (501) added TomcatDefault servlet
- (500) duplicate HttpServletRequest in classpath giving;
```
java.lang.NoSuchMethodError:
javax.servlet.http.HttpServletRequest.getHttpServletMapping()Ljavax/servlet/http/HttpServletMapping;
at org.apache.catalina.core.ApplicationHttpRequest.setRequest(ApplicationHttpRequest.java:690)
```
- Moved servlet-api to provided.
- Added missing deps.
- (404) Added workaround via extra profile to exclude servlet-api fully for tomcat
- Moved packaged version code to configsource bean.

 So it still does not work on tomcat, will give it later some more time.

johnament commented:
You know that Hammock already works with Webjars right?  If you add a dependency on a webjar to your package, it'll be available at `http://<host>:<port>/webjars/<artifact-id>/<artifact-version>`.  Is your goal so that its available at a different URL?

derekm commented on NOTICE:
```diff
@@ -6,6 +6,7 @@ John D. Ament
 Antoine Sabot-Durand
 Libor Kramoliš
 Derek P. Moore
+Willem Cazander
```
Yay!

derekm commented:
@immetoo -- I added Swagger UI to one of my projects using:
> https://github.com/hammock-project/hammock/wiki/Native-Filters#static-resources

by copying:
```
favicon-16x16.png
favicon-32x32.png
index.html
swagger-ui-bundle.js
swagger-ui-standalone-preset.js
swagger-ui.css
```
from `~/src/swagger-ui/dist/` to `~/src/[project]/src/main/resources/META-INF/resources/swagger-ui/`

The downside of including Swagger UI unmodified is: it spies on your schemas, sending all schemas loaded to SmartBear's remote "validation service"!

Schema spying can be turned off in swagger-ui's `index.html` by adding unsetting the `validatorUrl` config to Swagger UI's constructor:
```javascript
  const ui = SwaggerUIBundle({
    validatorUrl: null,
    url: "/openapi",
    ...
  })
```

Perhaps Hammock's Static Resources Native Filter could be improved to accept resource paths from MicroProfile Config! Then you could give it any random resources path out of which to serve static content. I don't know what happens when JAX-RS endpoints & static resources request the same paths, but that's already an issue with the existing Static Resources native filter, so people just need to be careful.

It should seem to me that if Hammock's serving of static content from resource paths was just slightly more powerful, then you could delete all the code you're trying to add in this PR.

Hope this helps!

derekm commented:
@johnament -- My QAs are starting to rely on the presence of Swagger UI in these new microservices I'm developing, so URLs of `http://localhost:8080/openapi` & `http://localhost:8080/swagger-ui` are much preferred, but currently my QAs have to remember  `http://localhost:8080/swagger-ui/index.html`, because Hammock's static resources native filter doesn't have anything like Apache HTTP Server's `DirectoryIndex` (https://httpd.apache.org/docs/2.4/mod/mod_dir.html#directoryindex).

Even so, remembering `index.html` is much preferable to them remembering Swagger UI version numbers as I intermittently upgrade those versions.

I vote for improving Hammock's Static Resources native filter until it can compete with Apache HTTP Server! 😄

johnament commented:
I wonder if you can solve this with a mix of what Hammock's doing right now and [OCP Rewrite](https://github.com/ocpsoft/rewrite).  I'll try to get a tutorial together if it works.

immetoo commented:
It Passes :)
Went for simple code, could not get the request dispatcher to work in tomcat.
Also could not use the ResourceManager as location is user definable which can stop loading webjars.
Have also added options for ui customization.

A module util-rewrite-ocp++ would be nice addition, but I rather not have it on my rest apis.

immetoo commented:
@derekm 
We try to let the build system do the coping, this webjar setup is already working for years.
Only then i did the customization in a local META-INF/resources/swagger-ui/index.html copy.
And changed manually all resources to versioned webjars urls.

Now this makes it all work by default without setup.
But like you suggested there are needs for raw index edits like for css or custom auth.
So have added custom index support and also enabled query parameter on it.
So validator can be disabled.

For your described setup the following settings should work;
```
swagger-ui.redirect=index.html?validatorUrl=
swagger-ui.path=/swagger-ui
swagger-ui.api=/openapi
```

immetoo commented:
Are there still open issue i can do here ? maybe integrate with ResourceManager more tests etc ?

johnament commented:
@immetoo Sorry! I actually thought I merged this earlier.
